---
title: "From mutations to Proteins: Omics Data Analysis Report"
author: "Mariangela Santorsola"
date: "03/01/2024"
output:
 html_document:
   toc: true
params:
    dds: rnaseq/deseq2.dds.RData
    vcf: resequencing/annotation/haplotypecaller/joint_variant_calling/joint_germline_recalibrated_snpEff.ann.vcf
---
# Result summary
The variant _<chr:start:ref:alt>_ has been prioritized for further validation assays. The variants occurrs in the _<up-|down-regulated ###>_ gene, and causes the protein mutation _<p.###>_ which is predicted to be _<###>_ by Aplhamissense. The variant is reported in _<ClinVar>_ as _<###>_, and gnomAD _<###>_.

# RNAseq data analysis

## Experimental design

```{r packages, eval=TRUE, echo=FALSE}
library(DESeq2)
library(tximport)
library(tidyverse)
library(pheatmap)
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
```

This section contains the results of a RNAseq dataset, including one control sample and one treatment sample, each in three biological replicates.

```{r data_input, eval=TRUE, echo=FALSE, include=TRUE}
groups <- rep(c("control", "case"), each = 3)
samples <-c("sample_01", "sample_02", "sample_03", "sample_04", "sample_05", "sample_06")
data <- tibble(Group = groups, Sample = samples) 
data %>% 
    knitr::kable()
```

## Data Processing

### Read alignment and Quantification

- The reads from FASTQ files were processed by **Salmon**, *a pseudo-alignment* tool included in the **nf-core/rnaseq** pipeline, using the Human Reference Genome GRCh38.

- The nf-core/rnaseq pipeline was lauched by the following command line:

```{bash, eval=FALSE, echo=TRUE, include=TRUE}
nextflow run nf-core/rnaseq -r 3.12.0 \
--input  datasets_lesson33/rnaseq/reads/rnaseq_samplesheet.csv \
--outdir results-in-you-bucket \
--genome GRCh38chr21 \
--pseudo_aligner salmon \
--skip_alignment \
--skip_biotype_qc \
-c rnaseq_nextflow.config \
-profile gls \
--skip_stringtie \
--skip_bigwig \
--skip_umi_extract \
--skip_trimming \
--skip_fastqc
```


- The config file used to launch nf-core/rnaseq is available at [rnaseq_nextflow.config](https://github.com/santorsola-teaching/class-lab-adv-omics-2023/blob/main/L34_from_mutations_to_proteins/rnaseq_nextflow.config)

## Differential Expression Analysis

- **DESeq2** R package was used to perform the Differential Expression (DE) analysis, and find the list of de-regulated genes when comparing **treaments vs controls**.

- The *gene-level* quantifications from nf-core/rnaseq were directly used to perform the DE analysis by DESeq2:

```{r differential_analysis, eval=TRUE, echo=TRUE, include=TRUE}

load(params$dds)

#define conditions, e.g. treatment and control
colData(dds)$Group1 <- as.factor(colData(dds)$Group1)

#define the design of comparison, treatments vs controls
design(dds) <- as.formula("~Group1")

#get Differentially Expressed genes
dds <- DESeq(dds)
```

### Summary

```{r print_results, eval=TRUE, echo=FALSE, include=TRUE}
res <- results(dds)
summary(res)
```

### MA plot

```{r maplot, eval=TRUE, echo=FALSE, include=TRUE, fig.cap="Plot showing the log2 fold change over the mean of normalized counts for analysed samples"}
plotMA(res, ylim=c(-3,3))
abline(h=c(-1,1), col="dodgerblue", lwd=2)
```

### Dispersion plot

```{r dispersion_plot, eval=TRUE, echo=FALSE, include=TRUE, fig.cap="Dispersion plot showing the final estimates shrunking from gene-wise estimates towards fitted estimates."}
plotDispEsts(dds)
```


### List of Differentially Expressed Genes

```{r significant_genes_table, eval=TRUE, echo=FALSE, include=TRUE}
resOrdered <- res[order(res$padj),]
resOrdered$Gene <- rownames(resOrdered)
sign_de_genes_table <- as_tibble(resOrdered[resOrdered$padj < 0.1,])
sign_de_genes_table %>% 
    dplyr::select(Gene, log2FoldChange, pvalue, padj) %>%
    knitr::kable()
significant_de_genes <- sign_de_genes_table$Gene
```
Table listing the statistically significan DE genes. For each gene, the log2 fold-change, p-value and adjusted pvalue are shown.


# Re-sequencing data analysis

## Experimental design

The re-sequencing dataset includes one control sample and one case sample.

```{r reseq_data_input, eval=TRUE, echo=FALSE, include=TRUE}
groups <- c("control", "case")
samples <-c("normal_1.000+disease_0.000", "normal_0.000+disease_1.000")
data <- tibble(Group = groups, Sample = samples) 
data %>% 
    knitr::kable()
```

## Data Processing

### Variant calling and annotation

-   The reads from FASTQ files were alygned by **BWA-mem**, the default alignment tool included in the **nf-core/sarek** pipeline, using the Human Reference Genome GRCh38.

-   **Haplotypecaller** was used to perform the variant calling.

-   **SnpEFF** was chosed to annotate the identified variants.

- The nf-core/rnaseq pipeline was lauched by using the following command line:

```{bash sarek_run, eval=FALSE, echo=TRUE, include=TRUE}
nextflow run nf-core/sarek -r 3.3.2 \
--input datasets_lesson33/germline/reads/sarek_samplesheet.csv \
--outdir results-in-you-bucket \
--igenomes_ignore true \
--genome GRCh38chr21 \
--tools haplotypecaller,snpeff \
--skip_tools haplotypecaller_filter \
--joint_germline \
--intervals datasets_lesson33/germline/chr21_intervals.list \
-c sarek_nextflow.config \
-profile gls
```

- The config file used to run nf-core/sarek is available at [sarek_nextflow.config](https://github.com/santorsola-teaching/class-lab-adv-omics-2023/blob/main/L34_from_mutations_to_proteins/sarek_nextflow.config)


```{r variant_settings, echo=FALSE, include=FALSE}
library(knitr)
library(VariantAnnotation)
library(tidyverse)
library(reticulate)
source("../extract_annotations.R")
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

Read vcf file from nf-core/sarek:

```{r import_VCF, eval=TRUE, echo=TRUE, include=TRUE, warning=FALSE, message=FALSE}
vcf <- readVcf(params$vcf)
```

```{r rowRanges, eval=TRUE, echo=FALSE, include=FALSE}
head(rowRanges(vcf))
```

```{r infoSection, eval=TRUE, echo=FALSE, include=FALSE}
head(info(vcf))
```

```{r genoSection, eval=TRUE, echo=FALSE, include=FALSE}
head(geno(vcf)$GT)
```

Create the table of annotations by the INFO field within the VCF file:

```{r simplifyVCF, message=FALSE, warning=FALSE}
variants <- rowRanges(vcf)
variants$`REF` <- as.character(variants$`REF`)
variants$`ALT` <- as.character(variants$`ALT`[[1]][[1]])
variants <- as_tibble(variants)
variants$variantName = names(rowRanges(vcf))
variants = cbind(variants, as_tibble(geno(vcf)$GT))
variants$gene <- unlist(lapply(info(vcf)$ANN, get_most_severe_gene))
variants$ens <- unlist(lapply(info(vcf)$ANN, get_most_severe_ens))
variants$transcript <- unlist(lapply(info(vcf)$ANN, get_most_severe_tr))
variants$aa_change <- unlist(lapply(info(vcf)$ANN, get_most_severe_aa_change))
variants$consequence <- unlist(lapply(info(vcf)$ANN, get_most_severe_consequence))
variants$impact <- unlist(lapply(info(vcf)$ANN, get_most_severe_impact))
```

### Localization and biological consequences of identified mutations

```{r consequences, eval=TRUE, echo=FALSE, include=TRUE, fig.cap="Distribution of the percentage of identified mutations across biological consequence categories"}
variants %>%
  filter(!is.na(consequence)) %>%
  count(consequence) %>%
  mutate(count = n,
         percent = paste0(round(count/sum(count) * 100, digits = 2), "%"))%>%
  arrange(desc(count)) %>%
  mutate(lab.ypos = cumsum(count) - 0.5 * count) %>%
  ggplot(aes(x="", y=count, fill=consequence))+
  geom_bar(width = 1, stat = "identity")+
  coord_polar(theta = "y")+
  ggrepel::geom_text_repel(aes(y=lab.ypos, label=percent), max.overlaps = Inf)+
  theme_void()
```

 


```{r consequences_table, eval=TRUE, echo=FALSE, include=TRUE}
variants %>%
  filter(!is.na(consequence)) %>%
  group_by(consequence) %>%
  summarise(count=n()) %>%
  arrange(desc(count)) %>%
  kable()
```

Total number of identified mutations across biological consequence categories within analysed samples.

## Variant Filtering

### Filtering Criteria

To filter the most clinically relevant mutations, the following process was followed:

1.  Select **mutations occurring in cases** and not present in matched controls
2.  Select the mutations with **HIGH** or **MODERATE** impact from SnpEff annotation
3.  Select the **mutations occurring in Differentially Expressed Genes**
4.  Select **missense** mutations
5.  Select mutations with **high Alphamissense pathogenicity score** 

```{r filterVariants, eval=TRUE, echo=TRUE, include=TRUE}
filteredVars = variants %>%
  filter(control_control1 == "0/0" & 
            (impact == "MODERATE" | impact == "HIGH") &
            ens %in% significant_de_genes & 
            consequence == "missense_variant") 
filteredVars <- as_tibble(filteredVars)
```

### Filtered mutations for further validation assays

```{r filtered_vars_table, eval=TRUE, echo=FALSE, include=TRUE}
selectedVars <- filteredVars %>%
  dplyr::select(seqnames, start, REF, ALT, gene, consequence, aa_change) 

selectedVars_table <- selectedVars 
selectedVars_table %>% knitr::kable()

```

### Alphamissense Pathogenicity Predictions of filtered variants

```{r alphamissense, eval=TRUE, echo=FALSE, include=TRUE}
selectedVars$alphamis_Pathogenicity = c(0.085, 0.149, 0.977, 0.062, 0.076, 0.106)
selectedVars$alphamis_Class = c("likely_benign", "likely_benign", "likely_pathogenic", "likely_benign", "likely_benign", "likely_benign")

selectedVars_table <- selectedVars 
selectedVars_table %>% knitr::kable()
```

### Gene constraint metrics from gnomAD


```{r gene_contraint, eval=TRUE, echo=FALSE, include=TRUE}
genes <- c("SOD1", "ETS2", "PSMG1")
loeuf <-c(1.515, 0.434, 0.927)
data <- tibble(Gene = genes, LOEUF = loeuf) 
data %>% 
    knitr::kable()
```


## Results: Prioritized variant(s)


```{r prioritized_variants, eval=TRUE, echo=FALSE, include=TRUE}
prioritised_variants <- selectedVars %>% filter(alphamis_Class == "likely_pathogenic")
prioritised_variants$Clinical_significance <- "Pathogenic/Likely pathogenic"
prioritised_variants$gnomAD_AF <- "variant not found"
prioritised_variants %>% 
    knitr::kable()
```
*_Clinical_significance_ annotations are from the ClinVar database.
